File: apps/packages/product-domain/src/workflows/interpolation.ts
/**
 * Pure template placeholder parsing + validation — format v2 (data-contract §1.3).
 *
 * Mirrors the server grammar (`server/.../workflows/domain/interpolation.py`):
 *
 *     {{inputs.<name>}}       -> a workflow input (eager at StartRun)
 *     {{<emit_name>.<field>}}  -> a field of an earlier `agent.emit` step
 *
 * Rules (identical to the server):
 *   - an input reference must name a declared input;
 *   - an emit reference must name an emit produced *strictly earlier* in run
 *     order (earlier nodes in full, earlier steps in the same node);
 *   - `inputs`, `steps`, `fields` are reserved first-segments (illegal emits).
 *
 * Pure and non-throwing for editor feedback: returns structured issues plus
 * match positions + autocomplete suggestions the panel editors consume.
 */

import { WORKFLOW_RESERVED_REF_SEGMENTS } from "./definition";

/** A placeholder that is `{{` (not backslash-escaped), a body, then `}}`. */
const PLACEHOLDER_RE = /(?<!\\)\{\{\s*([^{}]*?)\s*\}\}/g;
const INPUT_REF_RE = /^inputs\.([A-Za-z_][A-Za-z0-9_]*)$/;
const FIELDS_REF_RE = /^fields\.([A-Za-z_][A-Za-z0-9_]*)$/;
const EMIT_REF_RE = /^([A-Za-z_][A-Za-z0-9_]*)\.([A-Za-z_][A-Za-z0-9_]*)$/;

const RESERVED = new Set<string>(WORKFLOW_RESERVED_REF_SEGMENTS);

export type WorkflowReference =
  | { kind: "input"; name: string }
  | { kind: "emit"; emit: string; field: string }
  // Agent-filled notify field (track 3c). Legal only in a notify message that
  // declares agent_fields; the resolver rewrites it to an indexed ref.
  | { kind: "fields"; name: string };

export interface PlaceholderMatch {
  raw: string;
  body: string;
  reference: WorkflowReference | null;
  start: number;
  end: number;
}

export type TemplateReferenceCode =
  | "invalid_template_reference"
  | "unknown_input_reference"
  | "forward_emit_reference"
  | "fields_reference_not_allowed"
  | "unknown_field_reference";

export interface TemplateReferenceIssue {
  code: TemplateReferenceCode;
  message: string;
  match: PlaceholderMatch;
}

/** Parse a placeholder body into a typed reference, or null if malformed. */
export function parseReference(body: string): WorkflowReference | null {
  const inputMatch = INPUT_REF_RE.exec(body);
  if (inputMatch) {
    return { kind: "input", name: inputMatch[1]! };
  }
  // `fields.` before the generic emit rule (which would match `fields.name` then
  // fail the reserved-segment guard). Legality is enforced in validation.
  const fieldsMatch = FIELDS_REF_RE.exec(body);
  if (fieldsMatch) {
    return { kind: "fields", name: fieldsMatch[1]! };
  }
  const emitMatch = EMIT_REF_RE.exec(body);
  if (emitMatch && !RESERVED.has(emitMatch[1]!)) {
    return { kind: "emit", emit: emitMatch[1]!, field: emitMatch[2]! };
  }
  return null;
}

/** Every placeholder found in a string, in order, with positions. */
export function iterPlaceholders(value: string): PlaceholderMatch[] {
  const matches: PlaceholderMatch[] = [];
  PLACEHOLDER_RE.lastIndex = 0;
  let match: RegExpExecArray | null;
  while ((match = PLACEHOLDER_RE.exec(value)) !== null) {
    const raw = match[0];
    const body = (match[1] ?? "").trim();
    matches.push({
      raw,
      body,
      reference: parseReference(body),
      start: match.index,
      end: match.index + raw.length,
    });
  }
  return matches;
}

/** Every well-formed reference in a string, in order. */
export function iterReferences(value: string): WorkflowReference[] {
  const refs: WorkflowReference[] = [];
  for (const placeholder of iterPlaceholders(value)) {
    if (placeholder.reference) {
      refs.push(placeholder.reference);
    }
  }
  return refs;
}

/**
 * Validate every placeholder in one templated string field. `priorEmitNames` is
 * the set of emit names produced strictly before this field's step in run order.
 * Returns issues rather than throwing (editor feedback).
 */
export function validateStringReferences(
  value: string,
  options: {
    inputNames: ReadonlySet<string>;
    priorEmitNames: ReadonlySet<string>;
    /**
     * The agent_fields schema field names, supplied ONLY for a notify `message`
     * that declares agent_fields (track 3c). `null`/undefined = `{{fields.*}}` is
     * illegal in this context.
     */
    allowedFields?: ReadonlySet<string> | null;
  },
): TemplateReferenceIssue[] {
  const issues: TemplateReferenceIssue[] = [];
  const allowedFields = options.allowedFields ?? null;
  for (const placeholder of iterPlaceholders(value)) {
    const { reference } = placeholder;
    if (reference === null) {
      issues.push({
        code: "invalid_template_reference",
        message:
          `'${placeholder.raw}' is not a valid template reference `
          + "(expected {{inputs.NAME}} or {{EMIT.FIELD}}).",
        match: placeholder,
      });
      continue;
    }
    if (reference.kind === "input") {
      if (!options.inputNames.has(reference.name)) {
        issues.push({
          code: "unknown_input_reference",
          message: `Template references unknown input '${reference.name}'.`,
          match: placeholder,
        });
      }
    } else if (reference.kind === "fields") {
      if (allowedFields === null) {
        issues.push({
          code: "fields_reference_not_allowed",
          message:
            `'${placeholder.raw}' is only allowed in a notify step's message `
            + "that declares agent_fields.",
          match: placeholder,
        });
      } else if (!allowedFields.has(reference.name)) {
        issues.push({
          code: "unknown_field_reference",
          message:
            `Template references agent field '${reference.name}', which is not `
            + "declared in the notify step's agent_fields schema.",
          match: placeholder,
        });
      }
    } else if (!options.priorEmitNames.has(reference.emit)) {
      issues.push({
        code: "forward_emit_reference",
        message:
          `Template references emit '${reference.emit}', which is not produced `
          + "by an earlier step in run order.",
        match: placeholder,
      });
    }
  }
  return issues;
}

// --- Editor autocomplete -------------------------------------------------------

export type TemplateSuggestionKind = "input" | "emit";

export interface TemplateSuggestion {
  /** The token to insert, e.g. `{{inputs.repo}}`. */
  token: string;
  label: string;
  detail: string;
  kind: TemplateSuggestionKind;
}

export interface EmitSuggestionSource {
  /** The emit name. */
  emit: string;
  /** Human label for the producing step. */
  stepLabel: string;
  /** Fields the emit publishes (from its output_schema). */
  fieldNames: readonly string[];
}

export interface TemplateSuggestionInput {
  inputs: readonly { name: string; type: string }[];
  /** Emits produced by steps that run strictly before this one. */
  priorEmits: readonly EmitSuggestionSource[];
}

/**
 * The set of valid `{{...}}` tokens available at a given step: every declared
 * input, plus every field of a strictly-earlier emit.
 */
export function templateSuggestions(input: TemplateSuggestionInput): TemplateSuggestion[] {
  const suggestions: TemplateSuggestion[] = [];
  for (const inp of input.inputs) {
    suggestions.push({
      token: `{{inputs.${inp.name}}}`,
      label: `inputs.${inp.name}`,
      detail: inp.type,
      kind: "input",
    });
  }
  for (const source of input.priorEmits) {
    for (const field of source.fieldNames) {
      suggestions.push({
        token: `{{${source.emit}.${field}}}`,
        label: `${source.emit}.${field}`,
        detail: `from ${source.stepLabel}`,
        kind: "emit",
      });
    }
  }
  return suggestions;
}

File: server/proliferate/server/cloud/workflows/domain/interpolation.py
"""Pure template interpolation and input coercion for workflow plans (format v2).

Stored template grammar (data-contract §1.3 / B6): every templated string field
may contain

    {{inputs.<name>}}          -> a workflow input (eager, resolved at StartRun)
    {{<emit_name>.<field>}}     -> a field of an earlier ``agent.emit`` step

``StartRun`` resolves ``{{inputs.*}}`` **eagerly** (the values are known at run
creation) and **rewrites** ``{{<emit>.<field>}}`` to the runtime's indexed form
``{{steps[<n>].output.<field>}}`` at flatten time — so the runtime keeps its
existing indexed grammar (templates.rs is unchanged) and never learns emit names.

Reserved first segments (never legal emit names): ``inputs``, ``steps``,
``fields``. ``steps`` is reserved so the rewrite target can never collide with an
emit name; ``fields`` is reserved for the notify agent-filled follow-up.

Escaping: substitution is segment-based and never re-scanned. A substituted input
value additionally has every ``{`` / ``}`` backslash-escaped so a value that
literally contains ``{{steps[0].output.x}}`` can never survive resolution as a
live step-output token (injection guard). The runtime unescapes ``\\{`` / ``\\}``
back to literal braces after its own step-output pass.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from proliferate.constants.workflows import (
    WORKFLOW_INPUT_TYPE_BOOLEAN,
    WORKFLOW_INPUT_TYPE_CHOICE,
    WORKFLOW_INPUT_TYPE_NUMBER,
    WORKFLOW_INPUT_TYPE_TEXT,
    WORKFLOW_RESERVED_REF_SEGMENTS,
)

# A placeholder is ``{{`` not preceded by a backslash, then a reference, then
# ``}}``. References never contain braces, so ``[^{}]`` keeps matching greedy-safe.
_PLACEHOLDER_RE = re.compile(r"(?<!\\)\{\{\s*(?P<ref>[^{}]*?)\s*\}\}")
_INPUT_REF_RE = re.compile(r"^inputs\.(?P<name>[A-Za-z_][A-Za-z0-9_]*)$")
# {{<emit_name>.<field>}} — a two-segment ref whose first segment is not reserved.
_EMIT_REF_RE = re.compile(
    r"^(?P<emit>[A-Za-z_][A-Za-z0-9_]*)\.(?P<field>[A-Za-z_][A-Za-z0-9_]*)$"
)
# The runtime's indexed form — the *output* of the rewrite, kept parseable so a
# rewritten string still validates and re-emits verbatim.
_STEP_REF_RE = re.compile(r"^steps\[(?P<index>\d+)\]\.output\.(?P<name>[A-Za-z_][A-Za-z0-9_]*)$")


class TemplateReferenceError(Exception):
    """Raised when a template placeholder is malformed or resolves nowhere."""

    def __init__(self, code: str, message: str) -> None:
        self.code = code
        self.message = message
        super().__init__(message)


@dataclass(frozen=True)
class InputReference:
    name: str


@dataclass(frozen=True)
class EmitReference:
    emit: str
    field: str


@dataclass(frozen=True)
class StepOutputReference:
    index: int
    name: str


Reference = InputReference | EmitReference | StepOutputReference


def parse_reference(ref: str) -> Reference:
    """Parse a placeholder body into a typed reference, or raise."""

    input_match = _INPUT_REF_RE.match(ref)
    if input_match is not None:
        return InputReference(name=input_match.group("name"))
    step_match = _STEP_REF_RE.match(ref)
    if step_match is not None:
        return StepOutputReference(
            index=int(step_match.group("index")),
            name=step_match.group("name"),
        )
    emit_match = _EMIT_REF_RE.match(ref)
    if emit_match is not None:
        emit = emit_match.group("emit")
        if emit in WORKFLOW_RESERVED_REF_SEGMENTS:
            raise TemplateReferenceError(
                "invalid_template_reference",
                f"'{{{{{ref}}}}}' uses reserved first segment '{emit}'.",
            )
        return EmitReference(emit=emit, field=emit_match.group("field"))
    raise TemplateReferenceError(
        "invalid_template_reference",
        f"'{{{{{ref}}}}}' is not a valid template reference "
        "(expected {{inputs.NAME}} or {{EMIT.FIELD}}).",
    )


def iter_references(value: str) -> list[Reference]:
    """Return every typed reference found in a string, in order."""

    return [parse_reference(match.group("ref")) for match in _PLACEHOLDER_RE.finditer(value)]


def validate_string_references(
    value: str,
    *,
    input_names: frozenset[str],
    prior_emit_names: frozenset[str],
) -> None:
    """Validate every placeholder in one stored (pre-resolution) step string field.

    ``prior_emit_names`` is the set of ``agent.emit`` names that appear strictly
    before this field's step in run order (earlier spine nodes in full, earlier
    steps in the same node) — the visibility rule (data-contract §1.3). An emit
    ref must name one of them; an input ref must name a declared input. Indexed
    step-output refs are illegal in stored definitions (they are the resolver's
    output, not an authored form).
    """

    for reference in iter_references(value):
        if isinstance(reference, InputReference):
            if reference.name not in input_names:
                raise TemplateReferenceError(
                    "unknown_input_reference",
                    f"Template references unknown input '{reference.name}'.",
                )
        elif isinstance(reference, EmitReference):
            if reference.emit not in prior_emit_names:
                raise TemplateReferenceError(
                    "forward_emit_reference",
                    (
                        f"Template references emit '{reference.emit}', which is not "
                        "produced by an earlier step in run order."
                    ),
                )
        else:  # StepOutputReference
            raise TemplateReferenceError(
                "invalid_template_reference",
                "Indexed step-output references are not allowed in a stored definition; "
                "use {{EMIT.FIELD}}.",
            )


# --- Input coercion ------------------------------------------------------------


class ArgumentError(Exception):
    """Raised when provided run inputs do not satisfy the inputs schema."""

    def __init__(self, code: str, message: str) -> None:
        self.code = code
        self.message = message
        super().__init__(message)


@dataclass(frozen=True)
class ArgSpec:
    name: str
    type: str
    required: bool
    has_default: bool
    default: object
    enum_values: tuple[str, ...]


def _coerce_number(name: str, value: object) -> float | int:
    if isinstance(value, bool):
        raise ArgumentError("invalid_argument", f"Input '{name}' must be a number.")
    if isinstance(value, (int, float)):
        return value
    if isinstance(value, str):
        try:
            if value.strip().lstrip("+-").isdigit():
                return int(value)
            return float(value)
        except ValueError as exc:
            raise ArgumentError(
                "invalid_argument", f"Input '{name}' must be a number."
            ) from exc
    raise ArgumentError("invalid_argument", f"Input '{name}' must be a number.")


def _coerce_boolean(name: str, value: object) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, str) and value.strip().lower() in {"true", "false"}:
        return value.strip().lower() == "true"
    raise ArgumentError("invalid_argument", f"Input '{name}' must be a boolean.")


def _coerce_one(spec: ArgSpec, value: object) -> object:
    if spec.type == WORKFLOW_INPUT_TYPE_TEXT:
        if isinstance(value, (dict, list)):
            raise ArgumentError("invalid_argument", f"Input '{spec.name}' must be text.")
        return value if isinstance(value, str) else str(value)
    if spec.type == WORKFLOW_INPUT_TYPE_NUMBER:
        return _coerce_number(spec.name, value)
    if spec.type == WORKFLOW_INPUT_TYPE_BOOLEAN:
        return _coerce_boolean(spec.name, value)
    if spec.type == WORKFLOW_INPUT_TYPE_CHOICE:
        text = value if isinstance(value, str) else str(value)
        if text not in spec.enum_values:
            raise ArgumentError(
                "invalid_argument",
                f"Input '{spec.name}' must be one of {list(spec.enum_values)}.",
            )
        return text
    raise ArgumentError("invalid_argument", f"Input '{spec.name}' has an unknown type.")


def coerce_arguments(arg_specs: list[ArgSpec], provided: dict[str, object]) -> dict[str, object]:
    """Coerce and validate provided run inputs against the inputs schema.

    Rejects unknown inputs (strict), fills defaults, enforces required, and
    coerces each value to its declared type.
    """

    known = {spec.name for spec in arg_specs}
    unknown = set(provided) - known
    if unknown:
        raise ArgumentError(
            "unknown_argument",
            f"Unknown workflow input(s): {sorted(unknown)}.",
        )
    resolved: dict[str, object] = {}
    for spec in arg_specs:
        if spec.name in provided:
            resolved[spec.name] = _coerce_one(spec, provided[spec.name])
        elif spec.has_default:
            resolved[spec.name] = _coerce_one(spec, spec.default)
        elif spec.required:
            raise ArgumentError(
                "missing_argument",
                f"Required workflow input '{spec.name}' was not provided.",
            )
    return resolved


# --- Eager inputs interpolation + emit-ref rewrite -----------------------------


def _escape_braces(rendered: str) -> str:
    return rendered.replace("{", "\\{").replace("}", "\\}")


def _render_scalar(value: object) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


def resolve_string(
    value: str,
    *,
    inputs: dict[str, object],
    emit_index: dict[str, int],
) -> str:
    """Resolve one templated string at flatten time (the resolver's single pass).

    - ``{{inputs.*}}`` is interpolated eagerly (values brace-escaped so they can
      never introduce a new live token).
    - ``{{<emit>.<field>}}`` is rewritten to ``{{steps[<n>].output.<field>}}``
      using ``emit_index`` (emit name -> flattened step index). The runtime
      late-binds the indexed form.

    Segment-based (no re-scanning): replacements are concatenated as literal text.
    """

    out: list[str] = []
    last = 0
    for match in _PLACEHOLDER_RE.finditer(value):
        out.append(value[last : match.start()])
        reference = parse_reference(match.group("ref"))
        if isinstance(reference, InputReference):
            out.append(_escape_braces(_render_scalar(inputs[reference.name])))
        elif isinstance(reference, EmitReference):
            index = emit_index[reference.emit]
            out.append(f"{{{{steps[{index}].output.{reference.field}}}}}")
        else:  # already-indexed StepOutputReference — re-emit canonical
            out.append(f"{{{{steps[{reference.index}].output.{reference.name}}}}}")
        last = match.end()
    out.append(value[last:])
    return "".join(out)


def resolve_value(
    value: object,
    *,
    inputs: dict[str, object],
    emit_index: dict[str, int],
) -> object:
    """Recursively resolve every string within a JSON value (see ``resolve_string``)."""

    if isinstance(value, str):
        return resolve_string(value, inputs=inputs, emit_index=emit_index)
    if isinstance(value, list):
        return [resolve_value(item, inputs=inputs, emit_index=emit_index) for item in value]
    if isinstance(value, dict):
        return {
            key: resolve_value(item, inputs=inputs, emit_index=emit_index)
            for key, item in value.items()
        }
    return value
